SourceKit-LSP doesn't work with `swift-for-wasm-examples` package

Is there any way to configure the [`swift-for-wasm-examples`](https://github.com/apple/swift-for-wasm-examples) package so that development with [SourceKit-LSP](https://github.com/apple/sourcekit-lsp) works correctly?

When opening `LadderFilter.swift`, an error appears for `import VultDSP`:
```
Could not build C module 'VultDSP' SourceKit 
```

And in the errors I get

```
<wasi/api.h> only supports wasm32; doesn't yet support wasm64
...
```

In the output from SourceKit-LSP I can see: 
```
sourcekit: [1:getCursorInfo:426.8573] error creating ASTInvocation: warning: option '-incremental' is only supported in swift-driver
<unknown>:0: warning: option '-incremental' is only supported in swift-driver
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:2:10: note: in file included from <module-includes>:2:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h:19:17: error: unknown type name 'uint16_t'
static __inline uint16_t __bswap16(uint16_t __x)
                ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:2:10: note: in file included from <module-includes>:2:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h:19:36: error: unknown type name 'uint16_t'
static __inline uint16_t __bswap16(uint16_t __x)
                                   ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:2:10: note: in file included from <module-includes>:2:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h:24:17: error: unknown type name 'uint32_t'
static __inline uint32_t __bswap32(uint32_t __x)
                ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:2:10: note: in file included from <module-includes>:2:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h:24:36: error: unknown type name 'uint32_t'
static __inline uint32_t __bswap32(uint32_t __x)
                                   ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:2:10: note: in file included from <module-includes>:2:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h:29:17: error: unknown type name 'uint64_t'
static __inline uint64_t __bswap64(uint64_t __x)
                ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:2:10: note: in file included from <module-includes>:2:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/endian.h:29:36: error: unknown type name 'uint64_t'
static __inline uint64_t __bswap64(uint64_t __x)
                                   ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:27:2: error: <wasi/api.h> only supports wasm32; doesn't yet support wasm64
#error <wasi/api.h> only supports wasm32; doesn't yet support wasm64
 ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:31:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:31:
#include <stdint.h>
         ^
/home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/stdint.h:52:16: note: in file included from /home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/stdint.h:52:
# include_next <stdint.h>
               ^
/usr/include/stdint.h:26:10: note: in file included from /usr/include/stdint.h:26:
#include <bits/libc-header-start.h>
         ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:74:5: error: function-like macro '__GLIBC_USE' is not defined
#if __GLIBC_USE (IEC_60559_BFP_EXT) || __GLIBC_USE (ISOC2X)
    ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:31:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:31:
#include <stdint.h>
         ^
/home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/stdint.h:52:16: note: in file included from /home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/stdint.h:52:
# include_next <stdint.h>
               ^
/usr/include/stdint.h:26:10: note: in file included from /usr/include/stdint.h:26:
#include <bits/libc-header-start.h>
         ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:80:5: error: function-like macro '__GLIBC_USE' is not defined
#if __GLIBC_USE (IEC_60559_BFP_EXT) || defined __STDC_WANT_IEC_60559_EXT__
    ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:31:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:31:
#include <stdint.h>
         ^
/home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/stdint.h:52:16: note: in file included from /home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/stdint.h:52:
# include_next <stdint.h>
               ^
/usr/include/stdint.h:26:10: note: in file included from /usr/include/stdint.h:26:
#include <bits/libc-header-start.h>
         ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:97:5: error: function-like macro '__GLIBC_USE' is not defined
#if __GLIBC_USE (IEC_60559_FUNCS_EXT) || __GLIBC_USE (ISOC2X)
    ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:31:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:31:
#include <stdint.h>
         ^
/home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/stdint.h:52:16: note: in file included from /home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/stdint.h:52:
# include_next <stdint.h>
               ^
/usr/include/stdint.h:265:5: error: function-like macro '__GLIBC_USE' is not defined
#if __GLIBC_USE (IEC_60559_BFP_EXT_C2X)
    ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:41:16: error: static assertion failed due to requirement 'alignof(void *) == 4': non-wasi data layout
_Static_assert(_Alignof(void*) == 4, "non-wasi data layout");
               ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:41:32: note: expression evaluates to '8 == 4'
_Static_assert(_Alignof(void*) == 4, "non-wasi data layout");
                               ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:51:16: error: static assertion failed due to requirement 'sizeof(unsigned long) == 4': witx calculated size
_Static_assert(sizeof(__wasi_size_t) == 4, "witx calculated size");
               ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:51:38: note: expression evaluates to '8 == 4'
_Static_assert(sizeof(__wasi_size_t) == 4, "witx calculated size");
                                     ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:52:16: error: static assertion failed due to requirement 'alignof(unsigned long) == 4': witx calculated align
_Static_assert(_Alignof(__wasi_size_t) == 4, "witx calculated align");
               ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:52:40: note: expression evaluates to '8 == 4'
_Static_assert(_Alignof(__wasi_size_t) == 4, "witx calculated align");
                                       ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:689:16: error: static assertion failed due to requirement 'sizeof(__wasi_iovec_t) == 8': witx calculated size
_Static_assert(sizeof(__wasi_iovec_t) == 8, "witx calculated size");
               ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:689:39: note: expression evaluates to '16 == 8'
_Static_assert(sizeof(__wasi_iovec_t) == 8, "witx calculated size");
                                      ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:690:16: error: static assertion failed due to requirement 'alignof(__wasi_iovec_t) == 4': witx calculated align
_Static_assert(_Alignof(__wasi_iovec_t) == 4, "witx calculated align");
               ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:690:41: note: expression evaluates to '8 == 4'
_Static_assert(_Alignof(__wasi_iovec_t) == 4, "witx calculated align");
                                        ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:692:16: error: static assertion failed due to requirement '__builtin_offsetof(__wasi_iovec_t, buf_len) == 4': witx calculated offset
_Static_assert(offsetof(__wasi_iovec_t, buf_len) == 4, "witx calculated offset");
               ^
/home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/__stddef_offsetof.h:16:24: note: expanded from macro 'offsetof'
#define offsetof(t, d) __builtin_offsetof(t, d)
                       ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:692:50: note: expression evaluates to '8 == 4'
_Static_assert(offsetof(__wasi_iovec_t, buf_len) == 4, "witx calculated offset");
                                                 ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:710:16: error: static assertion failed due to requirement 'sizeof(__wasi_ciovec_t) == 8': witx calculated size
_Static_assert(sizeof(__wasi_ciovec_t) == 8, "witx calculated size");
               ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:710:40: note: expression evaluates to '16 == 8'
_Static_assert(sizeof(__wasi_ciovec_t) == 8, "witx calculated size");
                                       ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: note: while building module 'dlmalloc' imported from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:
#include <features.h>
         ^
<module-includes>:3:10: note: in file included from <module-includes>:3:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/errno.h:6:
#include <wasi_api.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:711:16: error: static assertion failed due to requirement 'alignof(__wasi_ciovec_t) == 4': witx calculated align
_Static_assert(_Alignof(__wasi_ciovec_t) == 4, "witx calculated align");
               ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/dlmalloc/include/wasi_api.h:711:42: note: expression evaluates to '8 == 4'
_Static_assert(_Alignof(__wasi_ciovec_t) == 4, "witx calculated align");
                                         ^
<unknown>:0: error: too many errors emitted, stopping now
<module-includes>:1:10: note: in file included from <module-includes>:1:
#include "/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/VultDSP/include/waveforms.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/VultDSP/include/waveforms.h:1:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/VultDSP/include/waveforms.h:1:
#include "../out.h"
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/VultDSP/include/../out.h:5:10: note: in file included from /home/bastian/Downloads/swift-for-wasm-examples-main/Sources/VultDSP/include/../out.h:5:
#include <stdint.h>
         ^
/home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/stdint.h:52:16: note: in file included from /home/bastian/Downloads/swift-DEVELOPMENT-SNAPSHOT-2024-05-15-a-ubuntu22.04/usr/lib/swift/clang/include/stdint.h:52:
# include_next <stdint.h>
               ^
/usr/include/stdint.h:26:10: note: in file included from /usr/include/stdint.h:26:
#include <bits/libc-header-start.h>
         ^
/usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33:10: error: could not build module 'dlmalloc'
#include <features.h>
         ^
/home/bastian/Downloads/swift-for-wasm-examples-main/Sources/swift-audio/Effects/LadderFilter.swift:13:8: error: could not build C module 'VultDSP'
import VultDSP
       ^
[org.swift.sourcekit-lsp:request-77] default 2024-05-19 20:53:59.8890 +0000
Succeeded (took 41.36097431182861ms)
```

It looks like the C flags in `Package.swift` are not used when building. 

I tried creating a `compile_flags.txt` with target and defines like in the `Package.swift`, but that doesn't seem to get respected.

--- a/Documentation/Configuration File.md
+++ b/Documentation/Configuration File.md
@@ -13,12 +13,15 @@ The structure of the file is currently not guaranteed to be stable. Options may 
 `config.json` is a JSON file with the following structure. All keys are optional and unknown keys are ignored.
 
 - `swiftPM`: Dictionary with the following keys, defining options for SwiftPM workspaces
-  - `configuration: "debug"|"release"`: The configuration to build the project for during background indexing and the configuration whose build folder should be used for Swift modules if background indexing is disabled
-  - `scratchPath: string`: Build artifacts directory path. If nil, the build system may choose a default value
-  - `cCompilerFlags: string[]`: Extra arguments passed to the compiler for C files
-  - `cxxCompilerFlags: string[]`: Extra arguments passed to the compiler for C++ files
-  - `swiftCompilerFlags: string[]`: Extra arguments passed to the compiler for Swift files
-  - `linkerFlags: string[]`: Extra arguments passed to the linker
+  - `configuration: "debug"|"release"`: The configuration to build the project for during background indexing and the configuration whose build folder should be used for Swift modules if background indexing is disabled. Equivalent to SwiftPM's `--configuration` option.
+  - `scratchPath: string`: Build artifacts directory path. If nil, the build system may choose a default value. Equivalent to SwiftPM's `--scratch-path` option.
+  - `swiftSDKsDirectory: string`: Equivalent to SwiftPM's `--swift-sdks-path` option
+  - `swiftSDK: string`: Equivalent to SwiftPM's `--swift-sdk` option
+  - `triple: string`: Equivalent to SwiftPM's `--triple` option
+  - `cCompilerFlags: string[]`: Extra arguments passed to the compiler for C files. Equivalent to SwiftPM's `-Xcc` option.
+  - `cxxCompilerFlags: string[]`: Extra arguments passed to the compiler for C++ files. Equivalent to SwiftPM's `-Xcxx` option.
+  - `swiftCompilerFlags: string[]`: Extra arguments passed to the compiler for Swift files. Equivalent to SwiftPM's `-Xswiftc` option.
+  - `linkerFlags: string[]`: Extra arguments passed to the linker. Equivalent to SwiftPM's `-Xlinker` option.
 - `compilationDatabase`: Dictionary with the following keys, defining options for workspaces with a compilation database
   - `searchPaths: string[]`: Additional paths to search for a compilation database, relative to a workspace root.
 - `fallbackBuildSystem`: Dictionary with the following keys, defining options for files that aren't managed by any build system

--- a/Sources/SKCore/SourceKitLSPOptions.swift
+++ b/Sources/SKCore/SourceKitLSPOptions.swift
@@ -23,19 +23,42 @@ import struct TSCBasic.AbsolutePath
 public struct SourceKitLSPOptions: Sendable, Codable {
   public struct SwiftPMOptions: Sendable, Codable {
     /// Build configuration (debug|release).
+    ///
+    /// Equivalent to SwiftPM's `--configuration` option.
     public var configuration: BuildConfiguration?
 
     /// Build artifacts directory path. If nil, the build system may choose a default value.
+    ///
+    /// Equivalent to SwiftPM's `--scratch-path` option.
     public var scratchPath: String?
 
+    /// Equivalent to SwiftPM's `--swift-sdks-path` option
+    public var swiftSDKsDirectory: String?
+
+    /// Equivalent to SwiftPM's `--swift-sdk` option
+    public var swiftSDK: String?
+
+    /// Equivalent to SwiftPM's `--triple` option
+    public var triple: String?
+
+    /// Equivalent to SwiftPM's `-Xcc` option
     public var cCompilerFlags: [String]?
+
+    /// Equivalent to SwiftPM's `-Xcxx` option
     public var cxxCompilerFlags: [String]?
+
+    /// Equivalent to SwiftPM's `-Xswiftc` option
     public var swiftCompilerFlags: [String]?
+
+    /// Equivalent to SwiftPM's `-Xlinker` option
     public var linkerFlags: [String]?
 
     public init(
       configuration: BuildConfiguration? = nil,
       scratchPath: String? = nil,
+      swiftSDKsDirectory: String? = nil,
+      swiftSDK: String? = nil,
+      triple: String? = nil,
       cCompilerFlags: [String]? = nil,
       cxxCompilerFlags: [String]? = nil,
       swiftCompilerFlags: [String]? = nil,
@@ -43,6 +66,9 @@ public struct SourceKitLSPOptions: Sendable, Codable {
     ) {
       self.configuration = configuration
       self.scratchPath = scratchPath
+      self.swiftSDKsDirectory = swiftSDKsDirectory
+      self.swiftSDK = swiftSDK
+      self.triple = triple
       self.cCompilerFlags = cCompilerFlags
       self.cxxCompilerFlags = cxxCompilerFlags
       self.swiftCompilerFlags = swiftCompilerFlags
@@ -53,6 +79,9 @@ public struct SourceKitLSPOptions: Sendable, Codable {
       return SwiftPMOptions(
         configuration: override?.configuration ?? base.configuration,
         scratchPath: override?.scratchPath ?? base.scratchPath,
+        swiftSDKsDirectory: override?.swiftSDKsDirectory ?? base.swiftSDKsDirectory,
+        swiftSDK: override?.swiftSDK ?? base.swiftSDK,
+        triple: override?.triple ?? base.triple,
         cCompilerFlags: override?.cCompilerFlags ?? base.cCompilerFlags,
         cxxCompilerFlags: override?.cxxCompilerFlags ?? base.cxxCompilerFlags,
         swiftCompilerFlags: override?.swiftCompilerFlags ?? base.swiftCompilerFlags,

--- a/Sources/SKSwiftPMWorkspace/SwiftPMBuildSystem.swift
+++ b/Sources/SKSwiftPMWorkspace/SwiftPMBuildSystem.swift
@@ -222,8 +222,29 @@ public actor SwiftPMBuildSystem {
       throw Error.cannotDetermineHostToolchain
     }
 
-    let swiftSDK = try SwiftSDK.hostSwiftSDK(AbsolutePath(destinationToolchainBinDir))
-    let swiftPMToolchain = try UserToolchain(swiftSDK: swiftSDK)
+    let hostSDK = try SwiftSDK.hostSwiftSDK(AbsolutePath(destinationToolchainBinDir))
+    let hostSwiftPMToolchain = try UserToolchain(swiftSDK: hostSDK)
+
+    var destinationSDK: SwiftSDK
+    if let swiftSDK = options.swiftSDK {
+      let bundleStore = try SwiftSDKBundleStore(
+        swiftSDKsDirectory: fileSystem.getSharedSwiftSDKsDirectory(
+          explicitDirectory: options.swiftSDKsDirectory.map { try AbsolutePath(validating: $0) }
+        ),
+        fileSystem: fileSystem,
+        observabilityScope: observabilitySystem.topScope,
+        outputHandler: { _ in }
+      )
+      destinationSDK = try bundleStore.selectBundle(matching: swiftSDK, hostTriple: hostSwiftPMToolchain.targetTriple)
+    } else {
+      destinationSDK = hostSDK
+    }
+
+    if let triple = options.triple {
+      destinationSDK = hostSDK
+      destinationSDK.targetTriple = try Triple(triple)
+    }
+    let destinationSwiftPMToolchain = try UserToolchain(swiftSDK: destinationSDK)
 
     var location = try Workspace.Location(
       forRootPackage: AbsolutePath(packageRoot),
@@ -244,7 +265,7 @@ public actor SwiftPMBuildSystem {
       fileSystem: fileSystem,
       location: location,
       configuration: configuration,
-      customHostToolchain: swiftPMToolchain
+      customHostToolchain: hostSwiftPMToolchain
     )
 
     let buildConfiguration: PackageModel.BuildConfiguration
@@ -265,20 +286,21 @@ public actor SwiftPMBuildSystem {
     self.toolsBuildParameters = try BuildParameters(
       destination: .host,
       dataPath: location.scratchDirectory.appending(
-        component: swiftPMToolchain.targetTriple.platformBuildPathComponent
+        component: hostSwiftPMToolchain.targetTriple.platformBuildPathComponent
       ),
       configuration: buildConfiguration,
-      toolchain: swiftPMToolchain,
+      toolchain: hostSwiftPMToolchain,
       flags: buildFlags
     )
 
     self.destinationBuildParameters = try BuildParameters(
       destination: .target,
       dataPath: location.scratchDirectory.appending(
-        component: swiftPMToolchain.targetTriple.platformBuildPathComponent
+        component: destinationSwiftPMToolchain.targetTriple.platformBuildPathComponent
       ),
       configuration: buildConfiguration,
-      toolchain: swiftPMToolchain,
+      toolchain: destinationSwiftPMToolchain,
+      triple: destinationSDK.targetTriple,
       flags: buildFlags
     )
 

--- a/Tests/SourceKitLSPTests/SwiftPMIntegration.swift
+++ b/Tests/SourceKitLSPTests/SwiftPMIntegration.swift
@@ -241,4 +241,53 @@ final class SwiftPMIntegrationTests: XCTestCase {
       ]
     )
   }
+
+  func testWasm() async throws {
+    let project = try await SwiftPMTestProject(
+      files: [
+        "/.sourcekit-lsp/config.json": """
+        {
+          "swiftPM": {
+            "triple": "wasm32-unknown-none-wasm"
+          }
+        }
+        """,
+        "Test.swift": """
+        #if arch(wasm32)
+        let _: UnsafeRawPointer = 1
+        #endif
+        """,
+      ],
+      manifest: """
+        let package = Package(
+          name: "WasmTest",
+          targets: [
+            .executableTarget(
+              name: "wasmTest",
+              cSettings: [.unsafeFlags(["-fdeclspec"])],
+              swiftSettings: [
+                .enableExperimentalFeature("Embedded"),
+                .interoperabilityMode(.Cxx),
+                .unsafeFlags(["-wmo", "-disable-cmo", "-Xfrontend", "-gnone"]),
+              ],
+              linkerSettings: [.unsafeFlags(["-Xclang-linker", "-nostdlib", "-Xlinker", "--no-entry"])]
+            )
+          ]
+        )
+        """
+    )
+
+    let (uri, _) = try project.openDocument("Test.swift")
+    let diagnostics = try await project.testClient.send(
+      DocumentDiagnosticsRequest(textDocument: TextDocumentIdentifier(uri))
+    )
+    guard case .full(let diagnostics) = diagnostics else {
+      XCTFail("Expected full diagnostics report")
+      return
+    }
+    XCTAssertEqual(
+      diagnostics.items.map(\.message),
+      ["Cannot convert value of type 'Int' to specified type 'UnsafeRawPointer'"]
+    )
+  }
 }

